src/mongo_client.ts

```diff
@@ -346,19 +346,29 @@ export class MongoClient extends EventEmitter implements OperationParent {
     const force = typeof forceOrCallback === 'boolean' ? forceOrCallback : false;
 
     return maybePromise(callback, cb => {
+      const completeClose = (err?: AnyError) => {
+        // clear out references to old topology
+        this.topology = undefined;
+        this.s.dbCache = new Map();
+        this.s.sessions = new Set();
+
```

Comment:
Per a discussion from #2355, we don't clean up `this.s.options` here, so we still get output warnings when re-connecting:
```
the options [servers] is not supported
the options [caseTranslate] is not supported
the options [dbName] is not supported
```

Reply:
I think this is fine for today. There's another unit of work here in the future (after the `MongoClientOptions` project) where the options will be stored immutably, in which case this problem goes away.